File: src/tools/shell.ts
import { z } from "zod";
import { execFile } from "child_process";
import type { ToolDefinition } from "./registry";
import { appConfig } from "../config";
import { resolveSafe } from "./file-utils";
import { detectShellInjection, truncateOutput } from "./sanitize";

/** Built-in blocked command patterns checked against the full command string before execution. */
const DEFAULT_COMMAND_BLOCKLIST = [
  "rm -rf /",
  "rm -rf /*",
  "mkfs",
  "dd if=",
  ":(){:|:&};:", // fork bomb
  "chmod -R 777 /",
  "chmod 777 /",
];

const schema = z.object({
  command: z.string().describe("Shell command to execute (split by whitespace; no shell expansion)"),
  cwd: z.string().optional().describe("Working directory (defaults to process.cwd())"),
  env: z.record(z.string()).optional().describe("Extra environment variables merged into the process environment"),
  timeout: z.number().optional().describe("Timeout in milliseconds (overrides TOOL_TIMEOUT_MS)"),
});

/** Structured result returned by the shell tool, serialised as JSON. */
interface ShellResult {
  stdout: string;
  stderr: string;
  exitCode: number;
}

/** Returns true if the command matches any pattern in the blocklist. */
function isBlocked(command: string, blocklist: string[]): boolean {
  const normalized = command.trim().replace(/\s+/g, " ");
  return blocklist.some((pattern) => normalized.includes(pattern));
}

export const toolDefinition: ToolDefinition = {
  name: "shell",
  description:
    "Execute a shell command via execFile (no shell injection risk). Returns stdout, stderr, and exit code as JSON. " +
    "The command string is split by whitespace to derive the executable and its arguments; " +
    "arguments that contain spaces are not supported through this field.",
  schema,
  permissions: "dangerous",
  execute: async ({
    command,
    cwd,
    env,
    timeout,
  }: {
    command: string;
    cwd?: string;
    env?: Record<string, string>;
    timeout?: number;
  }): Promise<string> => {
    // Merge built-in defaults with any user-configured blocked patterns
    const blocklist = [...DEFAULT_COMMAND_BLOCKLIST, ...appConfig.shellCommandBlocklist];

    if (isBlocked(command, blocklist)) {
      const result: ShellResult = {
        stdout: "",
        stderr: `Command blocked by blocklist: "${command}"`,
        exitCode: -1,
      };
      return JSON.stringify(result);
    }

    // Reject commands that contain shell injection metacharacters
    if (detectShellInjection(command)) {
      const result: ShellResult = {
        stdout: "",
        stderr: `Command blocked: shell injection metacharacters detected in "${command}"`,
        exitCode: -1,
      };
      return JSON.stringify(result);
    }

    if (!command.trim()) {
      return JSON.stringify({ stdout: "", stderr: "No command provided", exitCode: -1 } as ShellResult);
    }

    // Confine the working directory to the workspace root (path traversal prevention)
    let effectiveCwd: string;
    if (cwd) {
      try {
        effectiveCwd = resolveSafe(appConfig.workspaceRoot, cwd);
      } catch {
        return JSON.stringify({
          stdout: "",
          stderr: `Working directory "${cwd}" is outside the workspace root`,
          exitCode: -1,
        } as ShellResult);
      }
    } else {
      effectiveCwd = process.cwd();
    }

    const effectiveTimeout = timeout ?? appConfig.toolTimeoutMs;
    const isWindows = process.platform === "win32";

    // On Windows dispatch through cmd.exe so shell built-ins work.
    // On POSIX split by whitespace and pass directly to execFile (no shell spawned).
    const [executable, args]: [string, string[]] = isWindows
      ? ["cmd.exe", ["/c", command.trim()]]
      : (() => { const [exe, ...rest] = command.trim().split(/\s+/); return [exe, rest]; })()

    return new Promise<string>((resolve) => {
      execFile(
        executable,
        args,
        {
          cwd: effectiveCwd,
          env: { ...process.env, ...env },
          timeout: effectiveTimeout,
        },
        (error, stdout, stderr) => {
          if (error?.killed) {
            // Process was killed by the timeout mechanism
            resolve(
              JSON.stringify({
                stdout: "",
                stderr: `Command timed out after ${effectiveTimeout}ms`,
                exitCode: -1,
              } as ShellResult)
            );
            return;
          }

          // error.code is the numeric exit code when the process exits non-zero
          const exitCode =
            error === null ? 0 : typeof error.code === "number" ? error.code : 1;

          // Cap stdout and stderr to MAX_SHELL_OUTPUT_BYTES each
          const cappedStdout = truncateOutput(stdout, appConfig.maxShellOutputBytes);
          const cappedStderr = truncateOutput(stderr, appConfig.maxShellOutputBytes);

          resolve(JSON.stringify({ stdout: cappedStdout, stderr: cappedStderr, exitCode } as ShellResult));
        }
      );
    });
  },
};

File: src/agentloop/tools/shell.py
from __future__ import annotations

import asyncio
import os
import shlex

from pydantic import BaseModel, Field
from pydantic_ai import RunContext

from agentloop.agent import AgentDeps
from agentloop.tools._common import json_dumps, resolve_workspace_path
from agentloop.tools.registry import tool_def
from agentloop.tools.sanitize import detect_shell_injection, truncate_output

DEFAULT_COMMAND_BLOCKLIST = (
    "rm -rf /",
    "rm -rf /*",
    "mkfs",
    "dd if=",
    ":(){:|:&};:",
    "chmod -R 777 /",
    "chmod 777 /",
)


class ShellInput(BaseModel):
    command: str
    cwd: str | None = None
    env: dict[str, str] = Field(default_factory=dict)
    timeout: int | None = None


@tool_def(name="shell", description="Execute a shell command safely", permissions="dangerous")
async def shell(ctx: RunContext[AgentDeps], args: ShellInput) -> str:
    command = " ".join(args.command.strip().split())
    if not command:
        return json_dumps({"stdout": "", "stderr": "No command provided", "exit_code": -1})
    if any(pattern in command for pattern in (*DEFAULT_COMMAND_BLOCKLIST, *ctx.deps.settings.shell_command_blocklist)):
        return json_dumps({"stdout": "", "stderr": f'Command blocked by blocklist: "{command}"', "exit_code": -1})
    if detect_shell_injection(command):
        return json_dumps({
            "stdout": "",
            "stderr": f'Command blocked: shell injection metacharacters detected in "{command}"',
            "exit_code": -1,
        })
    cwd = resolve_workspace_path(ctx.deps, args.cwd or ".")
    proc = await asyncio.create_subprocess_exec(
        *shlex.split(command),
        cwd=str(cwd),
        env={**os.environ, **args.env},
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    timeout = (args.timeout or ctx.deps.settings.tool_timeout_ms) / 1000
    try:
        stdout_b, stderr_b = await asyncio.wait_for(proc.communicate(), timeout=timeout)
    except TimeoutError:
        proc.kill()
        await proc.communicate()
        return json_dumps({"stdout": "", "stderr": f"Command timed out after {int(timeout * 1000)}ms", "exit_code": -1})
    limit = ctx.deps.settings.max_shell_output_bytes
    return json_dumps({
        "stdout": truncate_output(stdout_b.decode(errors="ignore"), limit),
        "stderr": truncate_output(stderr_b.decode(errors="ignore"), limit),
        "exit_code": proc.returncode,
    })
